src/DotNetty.Handlers/Tls/TlsHandler.cs

```diff
@@ -28,42 +28,49 @@ public sealed class TlsHandler : ByteToMessageDecoder
         static readonly Action<Task, object> HandshakeCompletionCallback = new Action<Task, object>(HandleHandshakeCompleted);
 
         readonly SslStream sslStream;
+        readonly MediationStream mediationStream;
+        readonly TaskCompletionSource closeFuture;
+        readonly bool isServer;
+        readonly bool checkCertificateRevocation;
+        readonly SslProtocols enabledSslProtocols;
+        readonly X509Certificate2 serverCertificate;
+        readonly string targetHost;
+        readonly X509CertificateCollection clientCertificates;
+
         TlsHandlerState state;
         int packetLength;
-        readonly MediationStream mediationStream;
         volatile IChannelHandlerContext capturedContext;
         BatchingPendingWriteQueue pendingUnencryptedWrites;
         Task lastContextWriteTask;
-        readonly TaskCompletionSource closeFuture;
-        readonly bool isServer;
-        readonly X509Certificate2 certificate;
-        readonly string targetHost;
         bool firedChannelRead;
         IByteBuffer pendingSslStreamReadBuffer;
         Task<int> pendingSslStreamReadFuture;
 
-        TlsHandler(bool isServer, X509Certificate2 certificate, string targetHost, RemoteCertificateValidationCallback certificateValidationCallback)
+        public TlsHandler(Func<Stream, SslStream> sslStreamFactory, TlsSettings settings)
         {
-            Contract.Requires(!isServer || certificate != null);
-            Contract.Requires(isServer || !string.IsNullOrEmpty(targetHost));
+            Contract.Requires(sslStreamFactory != null);
+            Contract.Requires(settings != null);
 
             this.closeFuture = new TaskCompletionSource();
-
-            this.isServer = isServer;
-            this.certificate = certificate;
-            this.targetHost = targetHost;
+            this.enabledSslProtocols = settings.EnabledSslProtocols;
+            this.checkCertificateRevocation = settings.CheckCertificateRevocation;
+            var serverSettings = settings as ServerTlsSettings;
+            if (serverSettings != null)
+            {
+                this.isServer = true;
+                this.serverCertificate = serverSettings.Certificate;
+            }
+            else
+            {
+                var clientSettings = (ClientTlsSettings)settings;
+                this.isServer = false;
+                this.clientCertificates = clientSettings.Certificates;
+                this.targetHost = clientSettings.TargetHost;
+            }
             this.mediationStream = new MediationStream(this);
-            this.sslStream = new SslStream(this.mediationStream, true, certificateValidationCallback);
+            this.sslStream = sslStreamFactory(this.mediationStream);
         }
-
-        public static TlsHandler Client(string targetHost) => new TlsHandler(false, null, targetHost, null);
```

Comment:
should we leave these for common cases and internalize sslStreamFactory in such a case?